service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,298 @@
+		var deletedTimers, deletedUpdateInfos []string
+
+		var deletedSubStateMachine []*persistencepb.StateMachinePath
+		addTombstone := func(tombstone *persistencepb.StateMachineTombstone) error {
+			switch x := tombstone.StateMachineKey.(type) {
+			case *persistencepb.StateMachineTombstone_ActivityScheduledEventId:
+				deletedActivities = append(deletedActivities, x.ActivityScheduledEventId)
+			case *persistencepb.StateMachineTombstone_TimerId:
+				deletedTimers = append(deletedTimers, x.TimerId)
+			case *persistencepb.StateMachineTombstone_ChildExecutionInitiatedEventId:
+				deletedChildExecutionInfos = append(deletedChildExecutionInfos, x.ChildExecutionInitiatedEventId)
+			case *persistencepb.StateMachineTombstone_RequestCancelInitiatedEventId:
+				deletedRequestCancelInfos = append(deletedRequestCancelInfos, x.RequestCancelInitiatedEventId)
+			case *persistencepb.StateMachineTombstone_SignalExternalInitiatedEventId:
+				deletedSignalInfos = append(deletedSignalInfos, x.SignalExternalInitiatedEventId)
+			case *persistencepb.StateMachineTombstone_UpdateId:
+				deletedUpdateInfos = append(deletedUpdateInfos, x.UpdateId)
+			case *persistencepb.StateMachineTombstone_StateMachinePath:
+				deletedSubStateMachine = append(deletedSubStateMachine, x.StateMachinePath)
+			default:
+				return serviceerror.NewInvalidArgument("unknown tombstone type")
+			}
+			return nil
+		}
+		for _, tombstone := range tombstoneBatch {
+			for _, t := range tombstone.StateMachineTombstones {
+				if err := addTombstone(t); err != nil {
+					return nil, err
+				}
+			}
+		}
+		executionInfo := mu.GetExecutionInfo()
+		executionInfo.UpdateInfos = nil
+		executionInfo.SubStateMachinesByType = nil
+		mutation := &persistencepb.WorkflowMutableStateMutation{
+			UpdatedActivityInfos:       getUpdatedInfo(mu.GetPendingActivityInfos(), request.VersionedTransition),
+			UpdatedTimerInfos:          getUpdatedInfo(mu.GetPendingTimerInfos(), request.VersionedTransition),
+			UpdatedChildExecutionInfos: getUpdatedInfo(mu.GetPendingChildExecutionInfos(), request.VersionedTransition),
+			UpdatedRequestCancelInfos:  getUpdatedInfo(mu.GetPendingRequestCancelExternalInfos(), request.VersionedTransition),
+			UpdatedSignalInfos:         getUpdatedInfo(mu.GetPendingSignalExternalInfos(), request.VersionedTransition),
+			UpdatedUpdateInfos:         getUpdatedInfo(mu.GetExecutionInfo().UpdateInfos, request.VersionedTransition),
+			UpdatedSubStateMachines:    updateStateMachine,
+			DeletedActivities:          deletedActivities,
+			DeletedTimers:              deletedTimers,
+			DeletedChildExecutionInfos: deletedChildExecutionInfos,
+			DeletedRequestCancelInfos:  deletedRequestCancelInfos,
+			DeletedSignalInfos:         deletedSignalInfos,
+			DeletedUpdateInfos:         deletedUpdateInfos,
+			DeletedSubStateMachines:    deletedSubStateMachine,
+			SignalRequestedIds:         mu.GetPendingSignalRequestedIds(),
```

Comment:
There's also a LastUpdateVersionedTransition field in mutable state for signalRequestIDs, so we don't have sync those IDs every time.

Reply:
If this is the case, will this duplicated with the  `UpdatedSignalInfos` field?